libs/vkd3d/swapchain_new.c

```diff
@@ -0,0 +1,1785 @@
+        write_info.descriptorCount = 1;
+        image_info.imageView = chain->user.vk_image_views[chain->request.user_index];
+        image_info.imageLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+        image_info.sampler = VK_NULL_HANDLE;
+
+        VK_CALL(vkCmdPushDescriptorSetKHR(vk_cmd, VK_PIPELINE_BIND_POINT_GRAPHICS,
+                    chain->present.pipeline.vk_pipeline_layout, 0, 1, &write_info));
+
+        VK_CALL(vkCmdDraw(vk_cmd, 3, 1, 0, 0));
+    }
+
+    VK_CALL(vkCmdEndRenderingKHR(vk_cmd));
+
+    image_barrier.srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+    image_barrier.dstAccessMask = 0;
+    image_barrier.oldLayout = image_barrier.newLayout;
+    image_barrier.newLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR;
+
+    VK_CALL(vkCmdPipelineBarrier(vk_cmd,
+                VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
+                VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
+                0, 0, NULL, 0, NULL, 1, &image_barrier));
+}
+
+static bool dxgi_vk_swap_chain_submit_blit(struct dxgi_vk_swap_chain *chain, uint32_t swapchain_index)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &chain->queue->device->vk_procs;
+    VkDevice vk_device = chain->queue->device->vk_device;
+    VkSemaphoreCreateInfo semaphore_create_info;
+    VkCommandBufferAllocateInfo allocate_info;
+    VkCommandBufferBeginInfo cmd_begin_info;
+    VkFenceCreateInfo fence_create_info;
+    VkSubmitInfo submit_info;
+    VkCommandBuffer vk_cmd;
+    VkQueue vk_queue;
+    VkResult vr;
+
+    /* Create objects on-demand. */
+    if (!chain->present.vk_release_semaphores[swapchain_index])
+    {
+        semaphore_create_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO;
+        semaphore_create_info.pNext = NULL;
+        semaphore_create_info.flags = 0;
+        vr = VK_CALL(vkCreateSemaphore(vk_device, &semaphore_create_info,
+                    NULL, &chain->present.vk_release_semaphores[swapchain_index]));
+        if (vr < 0)
+        {
+            ERR("Failed to create semaphore, vr %d.\n", vr);
+            return false;
+        }
+    }
+
+    if (!chain->present.vk_blit_command_buffers[swapchain_index])
+    {
+        memset(&allocate_info, 0, sizeof(allocate_info));
+        allocate_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+        allocate_info.commandBufferCount = 1;
+        allocate_info.commandPool = chain->present.vk_blit_command_pool;
+        allocate_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+        VK_CALL(vkAllocateCommandBuffers(vk_device, &allocate_info,
+                    &chain->present.vk_blit_command_buffers[swapchain_index]));
+    }
+
+    if (chain->present.vk_blit_fences[swapchain_index])
+    {
+        vr = VK_CALL(vkWaitForFences(vk_device, 1, &chain->present.vk_blit_fences[swapchain_index], VK_TRUE, UINT64_MAX));
+        VKD3D_DEVICE_REPORT_BREADCRUMB_IF(chain->queue->device, vr == VK_ERROR_DEVICE_LOST);
+        if (vr < 0)
+            return false;
```

Comment:
this seems a tad inconsistent with error messages, we show one when semaphore creation above fails but not here or below in case fence creation goes wrong.

I guess this will only ever fail on `DEVICE_LOST`,  but might be worth priting anyway.